---
output:
  md_document:
    variant: markdown_github
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "show/README-",
  warning = FALSE
)
```


# omplotr: 'ggplot2' Based RNAseq Plot Function Collection

## Installation and loading

Install the latest version from [GitHub](https://github.com/bioShaun/omplotr) as follow: 
    
   
```{r, eval = FALSE}
# Install
if(!require(devtools)) install.packages("devtools")
devtools::install_github("bioShaun/omplotr")
```

## Theme

`theme_onmath` is a ggplot theme used in almost all rnaseq plots.


```{r theme_onmath, message=F, warning=F, fig.width = 6, fig.height = 5}
library(omplotr)
p <- ggplot(mtcars) + geom_point(aes(x = wt, y = mpg,colour = factor(gear)))
p + theme_onmath() + ggtitle("theme_onmath")
```

## Plot

functions to generate plot in ngs analysis

### QC

#### Reads GC distribution

```{r fastqc-gc, fig.width = 6, fig.height = 5}
# Fastqc GC result
head(gc_test_data, 4)

# lineplot of GC distribution across Fastq file
gc_line_plot(gc_test_data)
```

#### Reads Quality barplot

```{r fastqc-reads-quality, fig.width = 6, fig.height = 5}
# Reads Quality result
# Bars of Quality <= 30 were marked with color 'dodgerblue', 
# Bars of Quality > 30 were marked with color 'navy'.
head(rq_test_data, 4)

# Reads Quality barplot
reads_quality_plot(rq_test_data)
```

### Quant

#### expression box, violin and density plot

```{r exp-box, fig.width = 6, fig.height = 5}
# expression matrix
head(exp_test_data, 4)

# sample information
head(test_sample_data, 4)

# boxplot
om_boxplot(exp_test_data, test_sample_data, 'box')

# violin
om_boxplot(exp_test_data, test_sample_data, 'violin')

# density
om_boxplot(exp_test_data, test_sample_data, 'density')
```

```{r exp-box-merged, fig.width = 6, fig.height = 10}
# merged plot
om_boxplot(exp_test_data, test_sample_data, 'all')
```


#### expression PCA analysis point plot
```{r exp-pca, fig.width = 6, fig.height = 5}
om_pca_plot(exp_test_data, test_sample_data)
```

### expression correlation heatmap
```{r exp-cor}
om_correlation_plot(exp_test_data, test_sample_data)
```

#### diff expression volcano plot
```{r exp-volcano-single, fig.width = 6, fig.height = 8}
# diff result
head(diff_test_data, 4)

# plot volcano plot for a single compare
om_volcano_plot(diff_test_data, 'Case_vs_Control')
```

```{r exp-volcano-merge, fig.width = 6, fig.height = 12}
# plot volcano plot for merged results
om_volcano_plot(diff_test_data, 'ALL')
```

#### expression heatmap

```{r exp-heatmap, fig.width = 6, fig.height = 5}
# plot expression heatmap
om_heatmap(exp_test_data, test_sample_data)
```

#### expression cluster line plot
```{r exp-cluster, fig.width = 6, fig.height = 5}
# cluster result
head(cluster_test_data, 4)

# cluster plot
om_cluster_plot(cluster_test_data)
```

### Enrichment analysis

#### GO

```{r go, message=F, warning=F, fig.width = 2.5, fig.height = 2.5}
# diff genes
test_diff_genes <- go_test_data_list[['test_diff_genes']]
head(test_diff_genes, 4)

# gene length
test_gene_len <- go_test_data_list[['test_gene_len']]
head(test_gene_len, 4)

# get go annotation file
gene_go_map <- system.file("extdata", "topgo_test_data.txt", package = "omplotr")
gene_go_map_df <- data.table::fread(gene_go_map, header = F)
head(gene_go_map_df, 4)

# run goseq and show result
goseq_output <- om_goseq(test_diff_genes, test_gene_len, gene_go_map)
head(goseq_output, 4)

# run topGO
om_topgo(gene_go_map, test_diff_genes, goseq_output)
```

### Mapping

```{r bwa-mapping, fig.width = 6, fig.height = 5}

mapping_stats <- system.file("extdata", "all_sample.mapping.xls", package = "omplotr")

# show mapping stats table
mapping_df <- read.delim(mapping_stats)
head(mapping_df, 6)

# show mapping summary
bwa_mapping_plot(mapping_stats, 5)

# show detail sample information
bwa_mapping_plot(mapping_stats, 10)

```

