---
output:
  md_document:
    variant: markdown_github
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r}
knitr::opts_chunk$set(
  comment = "#>",
  fig.path = "show/README-",
  warning = FALSE,
  tidy = TRUE, 
  tidy.opts = list(comment = FALSE)
)
```


# omplotr: 'ggplot2' Based RNAseq Plot Function Collection

## Theme

`theme_onmath` is a ggplot theme used in almost all rnaseq plots.


```{r}
library(omplotr)
p <- ggplot(mtcars) + geom_point(aes(x = wt, y = mpg,colour = factor(gear)))
p + theme_onmath() + ggtitle("theme_onmath")
```

## Plot

functions to generate plot in ngs analysis

### QC

#### Reads GC distribution

```{r}
# Fastqc GC result
head(gc_test_data, 4)

# lineplot of GC distribution across Fastq file
gc_line_plot(gc_test_data)
```

#### Reads Quality barplot

```{r}
# Reads Quality result
# Bars of Quality <= 30 were marked with color 'dodgerblue', 
# Bars of Quality > 30 were marked with color 'navy'.
head(rq_test_data, 4)

# Reads Quality barplot
reads_quality_plot(rq_test_data)
```

### Quant

#### expression box, violin and density plot

```{r}
# expression matrix
head(exp_test_data, 4)

# sample information
head(test_sample_data, 4)

# boxplot
om_boxplot(exp_test_data, test_sample_data, 'box')

# violin
om_boxplot(exp_test_data, test_sample_data, 'violin')

# density
om_boxplot(exp_test_data, test_sample_data, 'density')
```

```{r, fig.width = 6, fig.height = 10}
# merged plot
om_boxplot(exp_test_data, test_sample_data, 'all')
```


#### expression PCA analysis point plot
```{r}
om_pca_plot(exp_test_data, test_sample_data)
```

### expression correlation heatmap
```{r}
om_correlation_plot(exp_test_data, test_sample_data)
```

#### diff expression volcano plot
```{r}
# diff result
head(diff_test_data, 4)

# plot volcano plot for a single compare
om_volcano_plot(diff_test_data, 'Case_vs_Control')

# plot volcano plot for merged results
om_volcano_plot(diff_test_data, 'ALL')
```

#### expression heatmap

```{r}
# plot expression heatmap
om_heatmap(exp_test_data, test_sample_data)
```

#### expression cluster line plot
```{r}
# cluster result
head(cluster_test_data, 4)

# cluster plot
om_cluster_plot(cluster_test_data)
```

### Enrichment analysis

#### GO

```{r}
# diff genes
head(go_test_data_list[['test_diff_genes']], 4)

# gene length
head(go_test_data_list[['test_gene_len']], 4)

# go annotation
head(go_test_data_list[['test_go_anno']], 4)

# run goseq and show result
goseq_output <- om_goseq(go_test_data_list[['test_diff_genes']], 
                         go_test_data_list[['test_gene_len']], 
                         go_test_data_list[['test_go_anno']])
head(goseq_output, 4)

# get topgo go annotation file
gene_go_map <- system.file("extdata", "topgo_test_data.txt", package = "omplotr")

# run topGO
run_topgo(gene_go_map, go_test_data_list[['test_diff_genes']], goseq_output)
```
