#' Plot bwa mapping stats
#'
#' Plot bwa mapping rate summrized by samtools
#'
#' Mapping rate of each sample is generated by \code{samtools stats sorted.bam};
#'
#' Out stats is extracted using grep and sed command:
#'
#' \code{grep ^SN | cut -f 2,3 | sed -n '3p;7p;8p;9p;10p;11p;13p;15p;16p;19p;20p;23p;24p;25p;31p;32p;38p;40p'};
#'
#' Stats of all samples in analysis were merged into a single table.
#'
#' @param mapping_stats merged mapping stats file
#' @param sample_limit  sample number limit to show detail inform of each sample
#' @param out_prefix    output file prefix, default is NULL, don't output file
#' @examples
#' mapping_stats <- system.file("extdata", "all_sample.mapping.xls", package = "omplotr")
#' # show mapping summary
#' bwa_mapping_plot(mapping_stats, 5)
#' # show detail sample information
#' bwa_mapping_plot(mapping_stats, 10)

bwa_mapping_plot <- function(mapping_stats, sample_limit, out_prefix=NULL) {

  mapping_plot_order <- c('reads mapped', 'reads mapped and paired',
                          'reads properly paired', 'reads MQ0',
                          'non-primary alignments', 'reads unmapped')

  mapping_df <- read.delim(mapping_stats)
  sample_number <- dim(mapping_df)[2] - 1
  mapping_df$Item <- stringr::str_replace(mapping_df$Item, ":", "")
  mapping_per_df <- mapping_df[, -1]
  mapped_reads <- as.numeric(mapping_per_df[1, ])
  mapping_per_df <- as.data.frame(t(as.matrix(t(mapping_per_df)) / mapped_reads))
  mapping_per_df$Item <- mapping_df$Item

  mapping_per_df <- dplyr::filter(mapping_per_df, Item %in% mapping_plot_order)
  mapping_per_df$Item <- factor(mapping_per_df$Item,
                                levels = mapping_plot_order)
  m_mapping_per_df <- reshape2::melt(mapping_per_df, id.vars = c('Item'))

  if (sample_number < sample_limit) {
    map_plot <- ggplot(m_mapping_per_df, aes(Item, value, fill=variable)) +
      geom_bar(stat = 'identity', position = position_dodge(),
               color='black')
  } else {
    map_plot <- ggplot(m_mapping_per_df, aes(Item, value, fill=Item)) +
      geom_boxplot()
  }

  if (sample_number <= 9) {
    sample_cols <- RColorBrewer::brewer.pal(sample_number, 'Set1')
  } else {
    sample_cols <- colorRampPalette(RColorBrewer::brewer.pal(9, 'Set1'))(sample_number)
  }
  map_plot <- map_plot +
    theme_onmath() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    scale_fill_manual(values = sample_cols) +
    xlab('') + ylab('Percentage of Reads') +
    guides(fill=guide_legend(title = 'SampleID')) +
    scale_y_continuous(labels = scales::percent)
  if (! is.null(out_prefix)) {
    out_dir <- dirname(out_prefix)
    path_name <- save_mkdir(out_dir)
    save_ggplot(ggplot_out = map_plot,
                output = out_prefix)
  }
  return(map_plot)
}

#' Plot Reads coverage distribution
#'
#' Plot bwa mapping coverage summrized by samtools
#'
#' Mapping coverage of each sample is generated by \code{samtools stats sorted.bam};
#'
#' Then extracted using \code{grep}
#'
#' \code{grep ^COV| cut -f 3,4}
#'
#' @param coverage_table coverage stats table of all samples
#' @param sample_limit   sample number limit to show detail inform of each sample
#' @param max_depth      Sequencing depth limit to show in plot
#' @param out_prefix     output file prefix, default is NULL, don't output file
#'
#' @examples
#'
#' cov_stats <- system.file("extdata", "all_sample.genome.cov.xls", package = "omplotr")
#'
#' # show coverage summary
#' reads_cov_plot(cov_stats, 5, 100)
#' # show detail sample information
#' reads_cov_plot(cov_stats, 10, 100)
reads_cov_plot <- function(coverage_table, sample_limit, max_depth, out_prefix=NULL) {
  cov.data <- read.table(coverage_table,header = T)
  cov_sum <- apply(cov.data[,c(-1)],2,sum)
  sample_num <- dim(cov.data)[2] -1
  cum.data <- cbind(cov.data$Depth,1-cumsum(cov.data[,c(-1)])/cov_sum)
  colnames(cum.data)[1] <- "Depth"
  cum_melt <- reshape2::melt(cum.data,id.vars = 1)
  colnames(cum_melt) <- c("Depth","SampleID","value")
  if (sample_num < sample_limit) {
    sample_col <- palette_colors(pal_name = 'Set1',
                                 sample_num = sample_num)
    plot_breaks <- c(c(0, max_depth/2, max_depth/10), max_depth)

    cov_fig <- ggplot(cum_melt, aes(Depth, value, colour = SampleID)) +
      geom_line(size = 1) +
      scale_x_continuous(breaks = plot_breaks,
                         limits = c(0, max_depth)) +
      scale_color_manual(values = sample_col)
  } else {
    plot_depth <- seq(max_depth/10, max_depth/2, max_depth/10)
    sel_cum_melt <- dplyr::filter(cum_melt, Depth %in% plot_depth)
    cov_fig <- ggplot(sel_cum_melt, aes(Depth, value, group=Depth, fill=Depth)) +
      geom_boxplot() +
      scale_fill_gradient(guide = F)
  }
  cov_fig <- cov_fig + ylab("Fraction of Reads \u2265 Depth") +
    theme_onmath()
  if (! is.null(out_prefix)) {
    out_dir <- dirname(out_prefix)
    path_name <- save_mkdir(out_dir)
    save_ggplot(ggplot_out = cov_fig,
                output = out_prefix)
  }
  return(cov_fig)
}

